cli/rt/11_url.js

```diff
@@ -876,6 +900,61 @@
     );
   }
 
+  function encodeStr(str, noEscapeTable, hexTable) {
+    const len = str.length;
+    if (len === 0) return "";
+
+    let out = "";
+    let lastPos = 0;
+
+    for (let i = 0; i < len; i++) {
+      let c = str.charCodeAt(i);
+      // ASCII
+      if (c < 0x80) {
+        if (noEscapeTable[c] === 1) continue;
+        if (lastPos < i) out += str.slice(lastPos, i);
+        lastPos = i + 1;
+        out += hexTable[c];
+        continue;
+      }
+
+      if (lastPos < i) out += str.slice(lastPos, i);
+
+      // Multi-byte characters ...
+      if (c < 0x800) {
+        lastPos = i + 1;
+        out += hexTable[0xc0 | (c >> 6)] + hexTable[0x80 | (c & 0x3f)];
+        continue;
+      }
+      if (c < 0xd800 || c >= 0xe000) {
+        lastPos = i + 1;
+        out += hexTable[0xe0 | (c >> 12)] +
+          hexTable[0x80 | ((c >> 6) & 0x3f)] +
+          hexTable[0x80 | (c & 0x3f)];
+        continue;
+      }
```

Comment:
This and other uses of `hexTable` look like a re-implementation of `encodeChar()`, which percent-encodes characters. Could you use that?

Reply:
this is great, did not know about `encodeChar`, will use this one